src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/Razor/ExtractToCodeBehindCodeActionResolver.cs

```diff
@@ -170,51 +173,63 @@ private static string GenerateCodeBehindPath(string path)
-    {
-        var mock = (ClassDeclarationSyntax)CSharpSyntaxFactory.ParseMemberDeclaration($"class Class {contents}")!;
-        var @class = CSharpSyntaxFactory
-            .ClassDeclaration(className)
-            .AddModifiers(CSharpSyntaxFactory.Token(CSharpSyntaxKind.PublicKeyword), CSharpSyntaxFactory.Token(CSharpSyntaxKind.PartialKeyword))
-            .AddMembers(mock.Members.ToArray())
-            .WithCloseBraceToken(mock.CloseBraceToken);
-
-        var @namespace = CSharpSyntaxFactory
-            .NamespaceDeclaration(CSharpSyntaxFactory.ParseName(namespaceName))
-            .AddMembers(@class);
-
-        var usings = FindUsings(razorCodeDocument)
-            .Select(u => CSharpSyntaxFactory.UsingDirective(CSharpSyntaxFactory.ParseName(u)))
-            .ToArray();
-        var compilationUnit = CSharpSyntaxFactory
-            .CompilationUnit()
-            .AddUsings(usings)
-            .AddMembers(@namespace);
-
-        return compilationUnit.NormalizeWhitespace().ToFullString();
+        var usingDirectives = razorCodeDocument
+                    .GetDocumentIntermediateNode()
+                    .FindDescendantNodes<UsingDirectiveIntermediateNode>();
+        foreach (var usingDirective in usingDirectives)
+        {
+            builder.Append("using ");
+
+            var content = usingDirective.Content;
+            var startIndex = content.StartsWith("global::", StringComparison.Ordinal)
+                ? 8
+                : 0;
+
+            builder.Append(content, startIndex, content.Length - startIndex);
+            builder.Append(';');
+            builder.AppendLine();
+        }
+
+        builder.AppendLine();
+        builder.Append("namespace ");
+        builder.AppendLine(namespaceName);
+        builder.Append('{');
+        builder.AppendLine();
+        builder.Append("public partial class ");
+        builder.AppendLine(className);
+        builder.AppendLine(contents);
+        builder.Append('}');
+
+        var newFileContent = builder.ToString();
+
+        var parameters = new FormatNewFileParams()
+        {
+            Project = new TextDocumentIdentifier
+            {
+                Uri = new Uri(project.FilePath, UriKind.Absolute)
+            },
+            Document = new TextDocumentIdentifier
+            {
+                Uri = codeBehindUri
+            },
+            Contents = newFileContent
+        };
+        var fixedContent = await _languageServer.SendRequestAsync<FormatNewFileParams, string?>(CustomMessageNames.RazorFormatNewFileEndpointName, parameters, cancellationToken).ConfigureAwait(false);
+
+        if (fixedContent is null)
+        {
+            // Sadly we can't use a "real" workspace here, because we don't have access. If we use our workspace, it wouldn't have the right settings
+            // for C# formatting, only Razor formatting, and we have no access to Roslyn's real workspace, since it could be in another process.
```

Comment:
Oh, this is weird. Thanks for adding the comment to explain. However, if we just pass an instance of `AdhocWorkspace`, I imagine that it's only ever going to get a default set of C# formatting options and will ignore anything the user might have set in VS Options or .editorconfig. Do you know what we need trom Roslyn to call into formatting without a workspace? I seem to remember that @CyrusNajmabadi was working on some sort of mechanism (`SolutionServices`?).

Reply:
Firstly, worth mentioning this path is now just a fallback until the Roslyn changes are inserted. Once that is in I wouldn't expect the above call to return null.

> Do you know what we need trom Roslyn to call into formatting without a workspace?

The `Format` method takes a `SyntaxFormattingOptions`, which you can see in the "real" call here: https://github.com/dotnet/roslyn/pull/69878/files#diff-7ab7cb7fc8b5d3ba31e8bbbdcd96a9dfcadc129bde4258aa73a91ab4b4720bfbR82
In theory we could call Roslyn to get that info, but if we're going to call Roslyn then the call on line 221 would have worked and we don't need this anyway :)